src/silo_api/database_watcher.cpp

```diff
@@ -0,0 +1,101 @@
+#include "silo_api/database_watcher.h"
+
+#include <ranges>
+
+#include <boost/algorithm/string/join.hpp>
+
+namespace {
+
+std::optional<silo::Database> findInitialDatabase(const std::filesystem::path& path) {
+   SPDLOG_INFO("Scanning path {} for valid data", path.string());
+   std::vector<std::pair<std::filesystem::path, silo::DataVersion>> all_found_data;
+   for (const auto& directory_entry : std::filesystem::directory_iterator{path}) {
+      SPDLOG_INFO("Checking directory entry {}", directory_entry.path().string());
+      if (!directory_entry.is_directory()) {
+         SPDLOG_INFO("Skipping {} because it is not a directory", directory_entry.path().string());
+         continue;
+      }
+      auto data_version = silo::DataVersion::fromString(directory_entry.path().filename().string());
+      if (data_version == std::nullopt) {
+         SPDLOG_INFO(
+            "Skipping {}. Its name is not a valid data version.", directory_entry.path().string()
+         );
+         continue;
+      }
+      SPDLOG_INFO(
+         "Found candidate data source {} with data version {}",
+         directory_entry.path().string(),
+         data_version->toString()
+      );
+      all_found_data.emplace_back(directory_entry.path(), std::move(*data_version));
+   }
+   if (all_found_data.empty()) {
+      SPDLOG_INFO("No previous data found, place data in {} for ingestion", path.string());
+      return std::nullopt;
+   }
+   if (all_found_data.size() == 1) {
+      const std::filesystem::path data_path = all_found_data.at(0).first;
+      SPDLOG_INFO("Using previous data: {}", data_path.string());
+      return silo::Database::loadDatabaseState(data_path);
+   }
+   auto max_element = std::max_element(
+      all_found_data.begin(),
+      all_found_data.end(),
+      [](const auto& element1, const auto& element2) { return element1.second < element2.second; }
+   );
+   SPDLOG_INFO(
+      "Selected database with highest data-version {} in path {} for ingestion",
+      max_element->second.toString(),
+      max_element->first.string()
+   );
+   return silo::Database::loadDatabaseState(max_element->first);
+}
+
+}  // namespace
+
+silo_api::DatabaseWatcher::DatabaseWatcher(const std::string& path, DatabaseMutex& database_mutex)
+    : watcher(path),
+      database_mutex(database_mutex) {
+   watcher.itemAdded += Poco::delegate(this, &DatabaseWatcher::onItemAdded);
+   auto initial_database = findInitialDatabase(path);
```

Comment:
I'm not sure whether I like this:
* Conceptually: Do we want to immediately start SILO with an empty database? We might need some way to tell consumers that it's currently loading data, but I think from an API perspective it would be nice if it would be available to answer calls as fast as possible.
* Codewise: This is quite unintuitive. This has the effect that calling a constructor blocks the program for a couple minutes until it loaded the data. Do we want to make this more explicit? Why does this have to happen in the constructor of a file watcher? We could move this to a background thread or simply call it in `api.cpp`:
```
  silo_api::DatabaseWatcher watcher(data_directory, database_mutex);
  auto initial_database = findInitialDatabase(path);
   if (initial_database.has_value()) {
      database_mutex.setDatabase(std::move(*initial_database));
   }
  ...
  server.start()
```

Reply:
I like your comments about the refactoring.

But what is the conceptual alternative?
If the database is started with a data directory that already contains data, I would find it rather intuitive to have it start with the newest data version it finds. Otherwise we need to always tell it "Hey, load the data in XY"

Reply:
Maybe doing it even before starting the DatabaseWatcher is even better? As we then do not have a race condition (albeit an almost impossible one), that between starting the watcher and looking for the initial database, a directory is added

```
  auto initial_database = findInitialDatabase(path);
   if (initial_database.has_value()) {
      database_mutex.setDatabase(std::move(*initial_database));
   }
  silo_api::DatabaseWatcher watcher(data_directory, database_mutex);
  ...
  server.start()
```

Reply:
I'm not sure what's better from a consumers perspective:
* Starting immediately with an empty database (might be confusing, since it's empty). Only after a few minutes, the database would be loaded and SILO shows data.
* Delaying the start until the database is loaded (might be confusing, since it takes a while until SILO responds at all)
* Immediately starting the API, but returning something like "sorry, still loading", until the initial DB is loaded.

Maybe the third option is the best?

Reply:
Maybe the alternative of removing the Watcher for directory events and instead using a timed event that checks all directory entries is better? Then we do not need any separate logic

Reply:
Might be, yes. It would probably use the same logic that we have for "find initial database", doesn't it? Since this requires the logic anyway, we already have it.

Reply:
Habe es angepasst! Habe den Timer zum starten auf 0 gesetzt, schaut also sofort nach einem Database State. Als Callback wird immer das directory durchsucht. (Manche log levels sollte man noch auf TRACE runtersetzen vermute ich). Sehr praktisch ist auch, dass ein Timer, solange ein Callback läuft, diesen nicht wieder aufruft. (Also wenn das loadDatabase länger als eine Intervallzeit läuft, ist direkt abgefangen )